hash_practice/exercises.py

```diff
@@ -22,8 +54,25 @@ def valid_sudoku(table):
         Each element can either be a ".", or a digit 1-9
         The same digit cannot appear twice or more in the same 
         row, column or 3x3 subgrid
-        Time Complexity: ?
-        Space Complexity: ?
-    """
-    pass
+        Time Complexity: o(n^2)
+        Space Complexity: O(n)

```

Comment:
I think your time complexity answer would be true if we accepted arbitrary sized grids of nxn, but luckily in this case we only accept 3x3 grids, so we can say this is O(1) for both time and space complexity.

However, your space complexity would be off in the nxn in case because we are also using up to n^2 elements in `cols`, `rows`, and `squares` (9 keys for each column/row/subgrid with up to 9 values each).

Regardless, this is an optional problem.